### PR TITLE
BUGFIX: Improve and fix type conversion

### PR DESCRIPTION
**Pull request by kdambekalns**
This fixes a bug when converting from array to `Resource`, base64-encoded data is
now decoded as it should be.

On top when converting to array, the there are now the properties `collectionName`, `relativePublicationPath`, `mediaType` and `sha1` contained in the resulting array.

### CODE REVIEW

**Comment by kitsunet** [comment]
👍 Looks good, I wonder why WIP

**Comment by kitsunet** [comment]
@kdambekalns what's still missing here?

**Comment by kdambekalns** [comment]
Well, this is part of a feature/package I started to work on, and I am pretty sure it will need more adjustments in that regard. If you deem this useful and fit for inclusion as is, fine with me. Even though it doesn't even have tests… ;)

File: TYPO3.Flow/Classes/TYPO3/Flow/ResourceManagement/ResourceTypeConverter.php [hunk 1]
```diff
@@ -231,7 +231,7 @@ protected function handleHashAndData(array $source, PropertyMappingConfiguration
         if ($resource === null) {
             $collectionName = $this->getCollectionName($source, $configuration);
             if (isset($source['data'])) {
-                $resource = $this->resourceManager->importResourceFromContent($source['data'], $source['filename'], $collectionName, $givenResourceIdentity);
+                $resource = $this->resourceManager->importResourceFromContent(base64_decode($source['data']), $source['filename'], $collectionName, $givenResourceIdentity);
```
Comment (kdambekalns) [inline TYPO3.Flow/Classes/TYPO3/Flow/ResourceManagement/ResourceTypeConverter.php]: This is an actual bugfix, btw…

**Comment by skurfuerst** [comment]
Actually, this change is for me a little hard to grasp; so I can't really vote on it.

**Comment by kdambekalns** [comment]
Merged master…

@skurfuerst hard to grasp? It adds more data to the converted representation, that's about it. :)

**Comment by albe** [comment]
I see how this is incomplete - the ResourceConverter should use the new properties (collectionName) and optionally validate the resource with the provided sha1. But since it contains a rather important bugfix I'd say merge it.

**Comment by kdambekalns** [comment]
@albe I am not sure I understand what you mean about the "new properties"…!?

**Comment by albe** [comment]
Sorry, mainly referring to the `'collectionName'` and `'sha1'` array keys, if you do (conceptually) `ResourceTypeConverter(ArrayConverter($resource))`.

**Comment by kdambekalns** [comment]
Ah, you mean to use them in the `ArrayConverter`, if present?

**Comment by albe** [comment]
In the `ResourceTypeConverter`, yep. Something like:

```
$collectionName = isset($source['collectionName']) ? $source['collectionName'] : $this->getCollectionName(...);
...
if (isset($source['sha1']) && $resource->getSha1() !== $source['sha1']) {
   throw new InvalidResourceDataException(...);
}
```

**Comment by kdambekalns** [comment]
@albe That should take care of your valid concerns, hopefully.

**Comment by albe** [comment]
Thx! Didn't even expect that to be resolved within this PR already :) :+1:

**Pull request merged**